lib/rules/no-useless-set-operand.ts

```diff
@@ -0,0 +1,242 @@
+import type { RegExpVisitor } from "@eslint-community/regexpp/visitor"
+import type {
+    CharacterClassElement,
+    ClassSetOperand,
+    ExpressionCharacterClass,
+    Node,
+    StringAlternative,
+} from "@eslint-community/regexpp/ast"
+import type { RegExpContext } from "../utils"
+import { createRule, defineRegexpVisitor } from "../utils"
+import { toUnicodeSet } from "regexp-ast-analysis"
+
+type FlatElement = CharacterClassElement | StringAlternative
+
+function getFlatElements(
+    node: ClassSetOperand | ExpressionCharacterClass["expression"],
+): readonly FlatElement[] {
+    if (node.type === "ClassStringDisjunction") {
+        return node.alternatives
+    }
+    if (node.type === "CharacterClass") {
+        const nested: FlatElement[] = []
+        // eslint-disable-next-line func-style -- x
+        const addElement = (element: CharacterClassElement) => {
+            if (element.type === "ClassStringDisjunction") {
+                nested.push(...element.alternatives)
+            } else if (element.type === "CharacterClass") {
+                if (!element.negate) {
+                    nested.push(...element.elements)
+                }
+                nested.push(element)
+            } else {
+                nested.push(element)
+            }
+        }
+        node.elements.forEach(addElement)
+        return nested
+    }
+
+    return []
+}
+
+function removeDescendant(root: Node, e: FlatElement): string {
+    let { start, end } = e
+
+    if (e.type === "StringAlternative") {
+        if (e.parent.alternatives.length === 1) {
+            // we have to remove the whole string disjunction
+            // eslint-disable-next-line no-param-reassign -- x
+            e = e.parent
+            start = e.start
+            end = e.end
+        } else {
+            // remove one adjacent | symbol
+            if (e.parent.alternatives.at(-1) === e) {
+                start--
+            } else {
+                end++
+            }
+        }
+    }
+
+    const before = root.raw.slice(0, start - root.start)
+    const after = root.raw.slice(end - root.start)
+    return before + after
+}
+
+export default createRule("no-useless-set-operand", {
+    meta: {
+        docs: {
+            description:
+                "disallow unnecessary elements in expression character classes",
+            category: "Best Practices",
+            // TODO Switch to recommended in the major version.
+            // recommended: true,
+            recommended: false,
```

Comment:
We are currently working on a major version update so we can include this rule in recommended ruleset.

```suggestion
            recommended: true,
```